src/Microsoft.AspNet.Routing/Template/TemplateParser.cs

```diff
@@ -134,22 +136,29 @@ private static bool ParseParameter(TemplateParserContext context, TemplateSegmen
 
             while (true)
             {
-                if (context.Current == Separator)
-                {
-                    // This is a dangling open-brace, which is not allowed
-                    context.Error = Resources.TemplateRoute_MismatchedParameter;
-                    return false;
-                }
-                else if (context.Current == OpenBrace)
+                if (context.Current == OpenBrace)
                 {
-                    // If we see a '{' while parsing a parameter name it's invalid. We'll just accept it for now
-                    // and let the validation code for the name find it.
+                    // This is an open brace inside of a parameter, it has to be escaped
+                    if (context.Next())
+                    {
+                        if (context.Current != OpenBrace)
+                        {
+                            context.Error = Resources.TemplateRoute_UnescapedBrace;
+                            return false;
+                        }
+                    }
+                    else
+                    {
+                        // This is a dangling open-brace, which is not allowed
+                        context.Error = Resources.TemplateRoute_MismatchedParameter;
```

Comment:
Consider with all of these things that you can include an example of what it catches. This would be something like: `{param{`. If you collect all the examples of the cases you're catching you can make sure they have something similar in tests. Obvi, I didn't follow my own advice when writing this code, but I wish I had.